// @framer/roray.bulma/code/FLC.tsx
import * as React from 'react';
import { PropertyControls, ControlType } from 'framer';
import { Field as _Field, Label as _Label, Control as _Control } from 'bloomer';

type Directions = 'left' | 'right';

// Define type of property
interface Props {
  //field
  fieldIsGrouped: boolean;
  fieldGroupDirection?: boolean | 'right' | 'centered';
  fieldHasAddons: boolean;
  fieldAddonsDirection?: boolean | 'right' | 'centered' | 'fullwidth';
  fieldIsHorizontal: boolean;

  //label
  labelIsSize: 'large' | 'medium' | 'small';
  labelText: string;

  //control
  controlIsLoading: boolean;
  contorlIsExpanded: boolean;
  controlHasIcons: boolean;
  controlIconDirections: Directions | Directions[];
}

export class FLC extends React.Component<Props> {
  // Set default properties
  static defaultProps = {};

  // Items shown in property panel
  static propertyControls: PropertyControls = {
    fieldIsGrouped: { type: ControlType.Boolean, title: 'Group Fields?' },
    fieldGroupDirection: {
      title: 'Field Group Direction',
      type: ControlType.SegmentedEnum,
      optionTitles: ['Right', 'Centered'],
      options: ['right', 'centered'],
      hidden(props) {
        return props.fieldIsGrouped === false;
      }
    },
    fieldHasAddons: { type: ControlType.Boolean, title: 'Field has addons?' },
    fieldAddonsDirection: {
      title: 'Field addons direction',
      type: ControlType.SegmentedEnum,
      optionTitles: ['Right', 'Centered', 'Full Width'],
      options: ['right', 'centered', 'fullwidth'],
      hidden(props) {
        return props.fieldHasAddons === false;
      }
    },
    fieldIsHorizontal: {
      type: ControlType.Boolean,
      title: 'Horizontal Field?'
    },
    labelIsSize: {
      title: 'Label Size',
      type: ControlType.SegmentedEnum,
      optionTitles: ['Large', 'Medium', 'Small'],
      options: ['large', 'medium', 'small']
    },
    labelText: { title: 'Label Text', type: ControlType.String },
    controlIsLoading: { type: ControlType.Boolean, title: 'Control Loading?' },
    contorlIsExpanded: {
      type: ControlType.Boolean,
      title: 'Control Expanded?'
    },
    controlHasIcons: { type: ControlType.Boolean, title: 'Control has icons?' }
  };

  render() {
    return (
      <_Field
        isGrouped={
          this.props.fieldIsGrouped ? this.props.fieldGroupDirection : false
        }
        hasAddons={
          this.props.fieldHasAddons ? this.props.fieldAddonsDirection : false
        }
        isHorizontal={this.props.fieldIsHorizontal}
      >
        <_Label isSize={this.props.labelIsSize}>{this.props.labelText}</_Label>
        <_Control
          isLoading={this.props.controlIsLoading}
          hasIcons={this.props.controlHasIcons}
          isExpanded={this.props.contorlIsExpanded}
        >
          {this.props.children}
        </_Control>
      </_Field>
    );
  }
}
