// @framer/roray.bulma/code/Notification.tsx
import * as React from 'react';
import { PropertyControls, ControlType } from 'framer';
import { Notification as _Notification } from 'bloomer';

const colors = {
  primary: 'primary',
  info: 'info',
  danger: 'danger',
  warning: 'warning',
  success: 'success',
  white: 'white',
  black: 'black',
  light: 'light',
  dark: 'dark',
  link: 'link'
};

// Define type of property
interface Props {
  isColor:
    | 'white'
    | 'light'
    | 'dark'
    | 'black'
    | 'primary'
    | 'info'
    | 'success'
    | 'warning'
    | 'danger';
  width: number;
  height: number;
}

export class Notification extends React.Component<Props> {
  // Set default properties
  static defaultProps = {
    isColor: 'primary'
  };

  // Items shown in property panel
  static propertyControls: PropertyControls = {
    isColor: {
      title: 'Color',
      type: ControlType.Enum,
      optionTitles: Object.keys(colors),
      options: Object.values(colors)
    }
  };

  render() {
    return (
      <_Notification
        {...this.props}
        style={{ width: this.props.width, height: this.props.height }}
      >
        {this.props.children}
      </_Notification>
    );
  }
}
