// @framer/roray.bulma/code/Progress.tsx
import * as React from 'react';
import { PropertyControls, ControlType } from 'framer';
import { Progress as _Progress } from 'bloomer';

const colors = {
  primary: 'primary',
  info: 'info',
  danger: 'danger',
  warning: 'warning',
  success: 'success',
  white: 'white',
  black: 'black',
  light: 'light',
  dark: 'dark',
  link: 'link'
};

const sizes = {
  large: 'large',
  medium: 'medium',
  small: 'small'
};

// Define type of property
interface Props {
  text: string;
  isColor:
    | 'white'
    | 'light'
    | 'dark'
    | 'black'
    | 'primary'
    | 'info'
    | 'success'
    | 'warning'
    | 'danger';
  isSize: 'large' | 'medium' | 'small';
  value: number;
  max: number;
}

export class Progress extends React.Component<Props> {
  // Set default properties
  static defaultProps = {
    text: 'Hello World!'
  };

  // Items shown in property panel
  static propertyControls: PropertyControls = {
    isColor: {
      title: 'Color',
      type: ControlType.Enum,
      optionTitles: Object.keys(colors),
      options: Object.values(colors)
    },
    isSize: {
      title: 'Size',
      type: ControlType.SegmentedEnum,
      optionTitles: Object.keys(sizes),
      options: Object.values(sizes)
    },
    value: {
      title: 'Value',
      type: ControlType.Number
    },
    max: {
      title: 'Max',
      type: ControlType.Number
    }
  };

  render() {
    return <_Progress {...this.props} />;
  }
}
