// @framer/roray.bulma/code/Table.tsx
import * as React from 'react';
import { PropertyControls, ControlType } from 'framer';
import { Table as _Table } from 'bloomer';

// Define type of property
interface Props {
  isBordered: boolean;
  isStriped: boolean;
  isNarrow: boolean;
}

export class Table extends React.Component<Props> {
  // Set default properties
  static defaultProps = { text: 'Hello World!' };

  // Items shown in property panel
  static propertyControls: PropertyControls = {
    isBordered: { title: 'Bordered', type: ControlType.Boolean },
    isStriped: { title: 'Striped', type: ControlType.Boolean },
    isNarrow: { title: 'Narrow', type: ControlType.Boolean }
  };

  render() {
    return (
      <_Table {...this.props}>
        <thead>
          <tr>
            <th>Name</th>
            <th>Score</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>Ryu</td>
            <td>10000</td>
          </tr>
          <tr>
            <td>Ken</td>
            <td>5000</td>
          </tr>
          <tr>
            <td>Akuma</td>
            <td>1200</td>
          </tr>
        </tbody>
      </_Table>
    );
  }
}
