// @framer/roray.bulma/code/TextArea.tsx
import * as React from 'react';
import { PropertyControls, ControlType } from 'framer';
import {
  Field as _Field,
  Label as _Label,
  Control as _Control,
  TextArea as _TextArea
} from 'bloomer';

const colors = {
  primary: 'primary',
  info: 'info',
  danger: 'danger',
  warning: 'warning',
  success: 'success',
  white: 'white',
  black: 'black',
  light: 'light',
  dark: 'dark',
  link: 'link'
};

const sizes = {
  large: 'large',
  medium: 'medium',
  small: 'small'
};

type Directions = 'left' | 'right';

// Define type of property
interface Props {
  //field
  fieldIsGrouped: boolean;
  fieldGroupDirection?: boolean | 'right' | 'centered';
  fieldHasAddons: boolean;
  fieldAddonsDirection?: boolean | 'right' | 'centered' | 'fullwidth';
  fieldIsHorizontal: boolean;

  //label
  labelIsSize: 'large' | 'medium' | 'small';
  labelText: string;

  //control
  controlIsLoading: boolean;
  contorlIsExpanded: boolean;
  controlHasIcons: boolean;
  controlIconDirections: Directions | Directions[];

  //Textarea
  textAreaIsSize: 'large' | 'medium' | 'small';
  textAreaIsActive: boolean;
  textAreaIsHovered: boolean;
  textAreaIsFocused: boolean;
  textAreaPlaceholder?: string;
  textAreaValue?: string;
}

export class TextArea extends React.Component<Props> {
  // Set default properties
  static defaultProps = {
    text: 'Hello World!'
  };

  // Items shown in property panel
  static propertyControls: PropertyControls = {
    fieldIsGrouped: { type: ControlType.Boolean, title: 'Group Fields?' },
    fieldGroupDirection: {
      title: 'Field Group Direction',
      type: ControlType.SegmentedEnum,
      optionTitles: ['Right', 'Centered'],
      options: ['right', 'centered'],
      hidden(props) {
        return props.fieldIsGrouped === false;
      }
    },
    fieldHasAddons: { type: ControlType.Boolean, title: 'Field has addons?' },
    fieldAddonsDirection: {
      title: 'Field addons direction',
      type: ControlType.SegmentedEnum,
      optionTitles: ['Right', 'Centered', 'Full Width'],
      options: ['right', 'centered', 'fullwidth'],
      hidden(props) {
        return props.fieldHasAddons === false;
      }
    },
    fieldIsHorizontal: {
      type: ControlType.Boolean,
      title: 'Horizontal Field?'
    },
    labelIsSize: {
      title: 'Label Size',
      type: ControlType.SegmentedEnum,
      optionTitles: ['Large', 'Medium', 'Small'],
      options: ['large', 'medium', 'small']
    },
    labelText: { title: 'Label Text', type: ControlType.String },
    controlIsLoading: { type: ControlType.Boolean, title: 'Control Loading?' },
    contorlIsExpanded: {
      type: ControlType.Boolean,
      title: 'Control Expanded?'
    },
    controlHasIcons: { type: ControlType.Boolean, title: 'Control has icons?' },
    textAreaIsSize: {
      title: 'textArea Size',
      type: ControlType.SegmentedEnum,
      optionTitles: ['Large', 'Medium', 'Small'],
      options: ['large', 'medium', 'small']
    },
    textAreaIsActive: { title: 'textArea Active?', type: ControlType.Boolean },
    textAreaIsHovered: {
      title: 'textArea Hovered?',
      type: ControlType.Boolean
    },
    textAreaIsFocused: {
      title: 'textArea Focused?',
      type: ControlType.Boolean
    },
    textAreaPlaceholder: {
      type: ControlType.String,
      title: 'textArea Placeholder'
    },
    textAreaValue: { type: ControlType.String, title: 'textArea Value' }
  };

  render() {
    return (
      <_Field
        isGrouped={
          this.props.fieldIsGrouped ? this.props.fieldGroupDirection : false
        }
        hasAddons={
          this.props.fieldHasAddons ? this.props.fieldAddonsDirection : false
        }
        isHorizontal={this.props.fieldIsHorizontal}
      >
        <_Label isSize={this.props.labelIsSize}>{this.props.labelText}</_Label>
        <_Control
          isLoading={this.props.controlIsLoading}
          hasIcons={this.props.controlHasIcons}
          isExpanded={this.props.contorlIsExpanded}
        >
          <_TextArea
            isSize={this.props.textAreaIsSize}
            isActive={this.props.textAreaIsActive}
            isHovered={this.props.textAreaIsHovered}
            isFocused={this.props.textAreaIsFocused}
            placeholder={this.props.textAreaPlaceholder}
            value={this.props.textAreaValue}
          />
        </_Control>
      </_Field>
    );
  }
}
