// @framer/markuli.materialui-theaming/code/MyCheckbox.tsx
import * as React from "react"
import { Frame } from "framer"
import CssBaseline from "@material-ui/core/CssBaseline"
import { addPropertyControls, ControlType } from "framer"
// tslint:disable-next-line: ban-ts-ignore
// @ts-ignore
import {
    makeStyles,
    withStyles,
    createMuiTheme,
    ThemeProvider,
} from "@material-ui/core/styles"
import MuiCheckbox from "@material-ui/core/Checkbox"
import Typography from "@material-ui/core/Typography"
// tslint:disable-next-line: ban-ts-ignore
// @ts-ignore
import FormControlLabel from "@material-ui/core/FormControlLabel"
import { myTheme } from "./MyTheme"
import { darkTheme } from "./MyDarkTheme"

interface Props {
    bgcolor?: "default" | "inherit" | "primary" | "secondary"
    checked?: boolean
    themeType?: boolean
    disabled?: boolean
    size?: string
    label?: string
    width?: number
    height?: number
    onChange?: (e: any) => void
}

const defaultProps: Props = {
    checked: false,
    themeType: false,
    bgcolor: "default",
    disabled: false,
    size: "medium",
    label: "Checkbox",
    width: 100,
    height: 42,
}

const useStyles = makeStyles({
    style: {
        "& .MuiCheckbox-root": {
            //color: "red",
            "& .MuiFormControlLabel-label": {
                //color: darkTheme.palette.text.primary,
                "&.Mui-disabled": {
                    //color: darkTheme.palette.text.disable,
                },
            },
        },
    },
})

export const MyCheckbox = (props: Props) => {
    const {
        checked: checkedProp,
        label,
        onChange,
        size,
        bgcolor,
        themeType,
        ...other
    } = props
    // tslint:disable-next-line: ban-ts-ignore
    // @ts-ignore

    const [checked, setChecked] = React.useState(false)

    const classes = useStyles()

    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        if (onChange) {
            onChange(event)
        }
        setChecked((event.target as HTMLInputElement).checked)
    }

    // tslint:disable-next-line: ban-ts-ignore
    // @ts-ignore
    React.useEffect(() => {
        setChecked(checkedProp)
    }, [checkedProp])

    const control = (
        <>
            <div
                style={{
                    opacity: checked ? 1 : 0,
                    borderRadius: "4px",
                    marginLeft: "12px",
                    position: "absolute",
                    backgroundColor: "black",
                    width: "16px",
                    height: "16px",
                }}
            ></div>
            <MuiCheckbox
                checked={checked}
                onChange={handleChange}
                size={size}
                color={bgcolor}
            ></MuiCheckbox>
        </>
    )

    return (
        <ThemeProvider theme={themeType ? darkTheme : myTheme}>
            <CssBaseline />
            <FormControlLabel
                //className={classes.style}
                control={control}
                label={label}
                {...other}
            />
        </ThemeProvider>
    )
}

MyCheckbox.defaultProps = defaultProps

addPropertyControls(MyCheckbox, {
    checked: {
        type: ControlType.Boolean,
        title: "Checked",
    },
    themeType: {
        type: ControlType.Boolean,
        title: "Dark theme",
    },
    bgcolor: {
        type: ControlType.Enum,
        title: "Color",
        options: ["default", "inherit", "primary", "secondary"],
    },
    disabled: {
        type: ControlType.Boolean,
        title: "Disabled",
    },
    size: {
        type: ControlType.Enum,
        title: "Size",
        options: ["small", "medium"],
    },
    label: {
        type: ControlType.String,
        title: "Label",
    },
})
