// @framer/markuli.materialui-theaming/code/MyDarkTheme.tsx
import * as React from "react"
import { addPropertyControls, ControlType } from "framer"
// tslint:disable-next-line: ban-ts-ignore
// @ts-ignore
import { colors } from "./canvas"
import { parseColor } from "./utils/parseColor"
import { MuiThemeProvider, createMuiTheme } from "@material-ui/core/styles"

const defaultTheme = createMuiTheme()

const myColors = {
    paletteType: "light",
    primary: colors.primary,
    primaryLight: colors.primaryLight,
    primaryDark: colors.primaryDark,
    primaryOrange: colors.primaryOrange,
    secondary: colors.secondary,
    default: colors.darkThemeDefault,
    defaultDark: colors.darkThemeDefaultDark,
    defaultLight: colors.darkThemeDefaultLight,
    error: "#f44336",
    info: "#2196f3",
    warning: "#ff9800",
    success: "#4caf4f",
}

export const darkTheme = createMuiTheme({
    //...defaultTheme,
    palette: {
        //...defaultTheme.palette,
        type: "dark",

        primary: {
            main: parseColor(myColors.primary),
            light: parseColor(myColors.primaryLight),
            dark: parseColor(myColors.primaryDark),
        },
        primaryOrange: parseColor(myColors.primaryOrange),

        secondary: { main: parseColor(myColors.secondary) },
        default: {
            main: parseColor(myColors.default),
            light: parseColor(myColors.defaultLight),
            dark: parseColor(myColors.defaultDark),
        },
        contrastThreshold: 3,
    },
    shadows: [
        "",
        "0 0 0 1px rgba(63,63,68,0.05), 0 1px 3px 0 rgba(63,63,68,0.15)",
    ],

    props: {
        // Name of the component ⚛️
        Card: {
            elevation: 2,
        },
        MuiButtonBase: {
            // The default props to change
            disableRipple: true, // No more ripple, on the whole application 💣!
            //disableElevation: true,
        },
        MuiTextField: {
            variant: "filled",
            label: "say something",
        },
    },
    typography: {
        fontFamily: [
            "Open Sans",
            "Roboto",
            '"Helvetica Neue"',
            "Arial",
            "sans-serif",
            '"Apple Color Emoji"',
            '"Segoe UI Emoji"',
            '"Segoe UI Symbol"',
        ].join(","),
    },
})
