// @framer/markuli.materialui-theaming/code/MyTheme.tsx
import * as React from "react"
import { addPropertyControls, ControlType } from "framer"
// tslint:disable-next-line: ban-ts-ignore
// @ts-ignore
import { colors } from "./canvas"
import { parseColor } from "./utils/parseColor"
import { MuiThemeProvider, createMuiTheme } from "@material-ui/core/styles"
//import { Test } from "./Test"

const myColors = {
    paletteType: "light",
    primary: colors.primary,
    primaryLight: colors.primaryLight,
    primaryDark: colors.primaryDark,
    secondary: colors.secondary,
    default: colors.default,
    defaultLight: colors.defaultLight,
    error: "#f44336",
    info: "#2196f3",
    warning: "#ff9800",
    success: "#4caf4f",
}

export const myTheme = createMuiTheme({
    palette: {
        //type: Test(),
        primary: {
            main: parseColor(myColors.primary),
            light: parseColor(myColors.primaryLight),
            dark: parseColor(myColors.primaryDark),
        },
        secondary: { main: parseColor(myColors.secondary) },
        default: {
            main: parseColor(myColors.default),
            light: parseColor(myColors.defaultLight),
        },
    },
    shadows: [
        "",
        "0 0 0 1px rgba(63,63,68,0.05), 0 1px 3px 0 rgba(63,63,68,0.15)",
    ],
    // overrides: {
    //     MuiInputBase: {
    //         root: {
    //             color: "black",
    //         },
    //     },
    //     MuiFilledInput: {
    //         root: {
    //             borderTopLeftRadius: 0,
    //         },
    //     },
    // },
    // props: {
    //     // Name of the component ⚛️
    //     Card: {
    //         elevation: 2,
    //     },
    //     MuiButtonBase: {
    //         // The default props to change
    //         disableRipple: true, // No more ripple, on the whole application 💣!
    //         disableElevation: true,
    //     },
    //     // MuiTextField: {
    //     //     variant: "filled",
    //     //     label: "say something",
    //     // },
    // },
    typography: {
        fontFamily: [
            "Open Sans",
            "BlinkMacSystemFont",
            '"Segoe UI"',
            "Roboto",
            '"Helvetica Neue"',
            "Arial",
            "sans-serif",
            '"Apple Color Emoji"',
            '"Segoe UI Emoji"',
            '"Segoe UI Symbol"',
        ].join(","),
    },
})
